src/soter/openssl/soter_ec_key_3.c

```diff
@@ -0,0 +1,439 @@
+        return sizeof(soter_ec_priv_key_384_t);
+    } else if (strcmp(curve, SN_secp521r1 /* P-521 */) == 0) {
+        return sizeof(soter_ec_priv_key_521_t);
+    } else {
+        return 0;
+    }
+}
+
+static char* ec_pub_key_tag(const char* curve)
+{
+    if (strcmp(curve, SN_X9_62_prime256v1 /* P-256 */) == 0) {
+        return EC_PUB_KEY_TAG(256);
+    } else if (strcmp(curve, SN_secp384r1 /* P-384 */) == 0) {
+        return EC_PUB_KEY_TAG(384);
+    } else if (strcmp(curve, SN_secp521r1 /* P-521 */) == 0) {
+        return EC_PUB_KEY_TAG(521);
+    } else {
+        return NULL;
+    }
+}
+
+static char* ec_priv_key_tag(const char* curve)
+{
+    if (strcmp(curve, SN_X9_62_prime256v1 /* P-256 */) == 0) {
+        return EC_PRIV_KEY_TAG(256);
+    } else if (strcmp(curve, SN_secp384r1 /* P-384 */) == 0) {
+        return EC_PRIV_KEY_TAG(384);
+    } else if (strcmp(curve, SN_secp521r1 /* P-521 */) == 0) {
+        return EC_PRIV_KEY_TAG(521);
+    } else {
+        return NULL;
+    }
+}
+
+static size_t bn_encode(const BIGNUM* bn, uint8_t* buffer, size_t length)
+{
+    int bn_size = BN_num_bytes(bn);
+    if (length < (size_t)bn_size) {
+        return 0;
+    }
+    memset(buffer, 0, length - bn_size);
+    return (length - bn_size) + BN_bn2bin(bn, buffer + (length - bn_size));
+}
+
+soter_status_t soter_engine_specific_to_ec_pub_key(const soter_engine_specific_ec_key_t* engine_key,
+                                                   bool compressed,
+                                                   soter_container_hdr_t* key,
+                                                   size_t* key_length)
+{
+    EVP_PKEY* pkey = (EVP_PKEY*)engine_key;
+    soter_status_t res;
+    size_t output_length;
+    char curve_str[16];
+    const char* param_name;
+    size_t serialized_len;
+
+    if ((!key_length) || (EVP_PKEY_EC != EVP_PKEY_id(pkey))) {
+        return SOTER_INVALID_PARAMETER;
+    }
+
+    // Curve identifier as string ("prime256v1", "secp384r1", "secp521r1" etc)
+    if (!EVP_PKEY_get_utf8_string_param(pkey,
+                                        OSSL_PKEY_PARAM_GROUP_NAME /* "group" */,
+                                        curve_str,
+                                        sizeof(curve_str),
+                                        NULL)) {
+        res = SOTER_INVALID_PARAMETER;
+        goto err;
+    }
+
+    if (!is_curve_supported(curve_str)) {
+        res = SOTER_INVALID_PARAMETER;
+        goto err;
+    }
+
+    output_length = ec_pub_key_size(curve_str, compressed);
+    if ((!key) || (output_length > *key_length)) {
+        *key_length = output_length;
+        res = SOTER_BUFFER_TOO_SMALL;
+        goto err;
+    }
+
+    *key_length = output_length;
+
+    if (compressed) {
+        param_name = OSSL_PKEY_PARAM_PUB_KEY /* "pub" */;
+
+        if (!EVP_PKEY_set_utf8_string_param(pkey, "point-format", "compressed")) {
+            res = SOTER_FAIL;
+            goto err;
+        }
```

Comment:
Hm... Does this work if you first try to get a compressed key, then uncompressed? (In either order.)

Or it works for Themis because it effectively allocates a new EVP_PKEY for every operation?

Reply:
It doesn't allocate new `EVP_PKEY*`, just switches something inside, which affects result of extracting `pub` property later.

Also, since I tried with two slightly different OpenSSL versions (3.0.7, 3.0.8), I decided to always extract non-compressed key from `encoded-pub-key` since there were some issues when only using this `pub` and `point-format` flag, don't remember the details but can find out, _probably the `point-format` property was not working in older version and `pub` always returned a compressed point_

Reply:
>  I decided to always extract non-compressed key from encoded-pub-key since there were some issues when only using this pub and point-format flag

Can you explain where you extract non-compressed keys? Just for note, previously themis returned only compressed keys. And some libraries/services may depend on that default behavior (acra-server for example, that relies on public key structure with compressed form). So, we should not change default behavior just because it is easier to do.

Reply:
> It doesn't allocate new EVP_PKEY*, just switches something inside

I mean the callers of this function. Themis never actually retains EVP_PKEY in memory long-term. It always imports key data into a newly created EVP_PKEY, then does stuff with it, then deallocates it and returns to the application. Or for key generation, creates a new EVP_PKEY with the key pair, exports it, and then deallocates. So accidentally we never use it more than once, and thus don't have a bug where you set the format in one operation and then it spookily applies to a different operation later.

Reply:
> since there were some issues when only using this `"pub"` and `"point-format"` flag, don't remember the details but can find out

Here's what [OpenSSL docs](https://www.openssl.org/docs/man3.0/man7/EVP_PKEY-EC.html) have to say about `"pub"`:

> "pub" (**OSSL_PKEY_PARAM_PUB_KEY**) <octet string>
>
> [...]
>
> Note, in particular, that the choice of point compression format used for encoding the exported value via EVP_PKEY_todata() depends on the underlying provider implementation. Before OpenSSL 3.0.8, the implementation of providers included with OpenSSL always opted for an encoding in compressed format, unconditionally. Since OpenSSL 3.0.8, the implementation has been changed to honor the **OSSL_PKEY_PARAM_EC_POINT_CONVERSION_FORMAT** parameter [`"point-format"`], if set, or to default to uncompressed format.

I'd just say, "you need to use OpenSSL 3.0.8+". Then always pass the format explicitly.

Uncompressed keys are a new feature in Themis 0.15. OpenSSL 3.0 support is also a new feature as well, never guaranteed to work in 0.14. We can add some restrictions, relax them later for users that absolutely must use outdated OpenSSL.

Reply:
In `3.0.7` you could get uncompressed key from `encoded-pub-key` param. At least, that's what worked for me, in both this version and in `3.0.8`. That's why I decided to leave this method for uncompressed key for all versions.

More important question: what about OS support? Cutting everything below `3.0.8` could drop some distributions where slightly older OpenSSL is installed and frozen. One such example is [Ubuntu 22.04 using 3.0.2](https://packages.ubuntu.com/search?keywords=openssl&searchon=names&suite=jammy&section=all). Meanwhile, Debian seem to use [3.0.8 in bookworm (testing)](https://packages.debian.org/search?keywords=openssl&searchon=names&suite=testing&section=all) and [1.1.1 in bullseye (stable)](https://packages.debian.org/search?keywords=openssl&searchon=names&suite=stable&section=all).

Also, I may try adding some macro switch to not set `point-format` on `openssl <3.0.8` and just read `pub` with compressed-by-default key there.

Reply:
Uh... I didn't mean do not support OpenSSL 3.0.0–3.0.7 completely, rather do not support uncompressed keys there. Just to reduce the amount of hacks and compatibility shims for versions that will become obsolete the next year.

Reply:
So you mean
- read `pub` in `<3.0.8` for compressed keys
- return error in `<3.0.8` for uncompressed
- write `point-format` and read `pub` in `3.0.8` for both key kinds

?

Reply:
Something like that. Just always do

> write point-format and read pub in 3.0.8 for both key kinds

Then 3.0.8+ will return what we expect and 3.0.0–3.0.7 will ignore the requested format and always return compressed key, as if uncompressed key feature does not exist.

Don't know even it's worth to do this:

> return error in <3.0.8 for uncompressed

But it looks like we need to do work anyway. Either document that it does not work on Ubuntu 22.04 and won't work in the nearest future. Or add an explicit version check to return THEMIS_NOT_SUPPORTED for this case, making it more clear that uncompressed keys do not work. Or add some hacks to support them on Ubuntu 22.04 if it's important.

Reply:
btw, regarding

> write point-format

Could it be that you don't need to set it on EVP_PKEY itself via EVP_PKEY_set_utf8_string_param(), but rather use EVP_PKEY_todata() and specify the parameter just for that call?

Reply:
Never actually tried `todata()` API. But it may turn out to be a better fit :thinking: